File: src/main/python/AoC2019_02.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 2
#

import itertools
import sys
from copy import deepcopy

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.intcode import IntCode

Input = list[int]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return IntCode.parse(list(input_data))

    def run_program(self, prog: list[int], noun: int, verb: int) -> int:
        prog[1] = noun
        prog[2] = verb
        int_code = IntCode(prog)
        int_code.run()
        return int_code.get_program()[0]

    def part_1(self, program: list[int]) -> int:
        return self.run_program(program, 12, 2)

    def part_2(self, program: list[int]) -> int:
        for noun, verb in itertools.product(range(100), repeat=2):
            if self.run_program(deepcopy(program), noun, verb) == 19_690_720:
                return 100 * noun + verb
        raise RuntimeError("Unsolved")

    def samples(self) -> None:
        pass


solution = Solution(2019, 2)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2019_02.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static com.github.pareronia.aoc.IterTools.product;

import java.util.ArrayList;
import java.util.List;

import com.github.pareronia.aoc.intcode.IntCode;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2019_02 extends SolutionBase<List<Long>, Long, Integer> {
    
    private AoC2019_02(final boolean debug) {
        super(debug);
    }

    public static AoC2019_02 create() {
        return new AoC2019_02(false);
    }

    public static AoC2019_02 createDebug() {
        return new AoC2019_02(true);
    }
    
    private Long runProgram(
            final List<Long> program,
            final long noun,
            final long verb
    ) {
        final List<Long> theProgram = new ArrayList<>(program);
        theProgram.set(1, noun);
        theProgram.set(2, verb);
        final IntCode intCode = new IntCode(theProgram, this.debug);
        intCode.run(theProgram);
        return intCode.getProgram().get(0);
    }

    @Override
    protected List<Long> parseInput(final List<String> inputs) {
        return IntCode.parse(inputs.get(0));
    }

    @Override
    public Long solvePart1(final List<Long> program) {
        return runProgram(program, 12, 2);
    }

    @Override
    public Integer solvePart2(final List<Long> program) {
        return product(range(100), range(100)).stream()
            .filter(p -> runProgram(program, p.get(0), p.get(1)) == 19_690_720)
            .map(p -> 100 * p.get(0) + p.get(1))
            .findFirst().orElseThrow();
    }

    public static void main(final String[] args) throws Exception {
        AoC2019_02.create().run();
    }
}